src/components/Popover/Popover.tsx

```diff
@@ -0,0 +1,247 @@
+     */
+    onOpen?: () => void;
+    /**
+     * Optional: Define a callback for when Popover content is closed
+     */
+    onClose?: () => void;
+}
+
+export const Popover: React.FC<PopoverProps> = ({
+    children,
+    content = '',
+    placement = 'bottom-start',
+    offset = 5,
+    isOpen = false,
+    onOpen,
+    onClose
+}: PopoverProps) => {
+    const [triggerReference, setTriggerReference] = React.useState(undefined);
+    const [popperReference, setPopperReference] = React.useState(undefined);
+    const popoverTriggerRef = React.useRef<HTMLDivElement>(null);
+    const popoverContentRef = React.useRef<HTMLDivElement>(null);
+
+    const [openByDefault, setOpenByDefault] = React.useState(isOpen);
+
+    const [render, setRender] = React.useState(openByDefault);
+
+    const { styles, attributes } = usePopper(triggerReference, popperReference, {
+        placement,
+        strategy: 'fixed',
+        modifiers: [
+            {
+                name: 'offset',
+                enabled: !!offset,
+                options: {
+                    offset: [0, offset]
+                }
+            },
+            {
+                name: 'flip',
+                enabled: true
+            }
+        ]
+    });
+
+    const resolveCallback = React.useCallback(
+        state => {
+            if (onClose && !state) onClose();
+            if (onOpen && state) onOpen();
+        },
+        [onClose, onOpen]
+    );
+
+    const hidePopover: () => void = React.useCallback(() => {
+        if (openByDefault) {
+            setOpenByDefault(false);
+        } else {
+            setRender(false);
+        }
+        resolveCallback(false);
+    }, [openByDefault, resolveCallback]);
+
+    const handleClose = React.useCallback(() => {
+        if (render) {
+            hidePopover();
+        }
+    }, [render, hidePopover]);
+
+    const handleClick: () => void = React.useCallback(() => {
+        if (render) {
+            hidePopover();
+        } else {
+            setRender(true);
+            resolveCallback(true);
+        }
+    }, [resolveCallback, setRender, render, hidePopover]);
+
+    // handleOut - handles click outside the trigger
+    const handleOut = React.useCallback(
+        ev => {
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+            if (popoverTriggerRef && popoverTriggerRef.current && !popoverTriggerRef.current.contains(ev.target)) {
+                if (!openByDefault) {
+                    handleClose();
+                } else {
+                    setOpenByDefault(false);
+                    resolveCallback(false);
+                }
+            }
+        },
+        [openByDefault, popoverTriggerRef, handleClose, resolveCallback]
+    );
+
+    const handleKeyControl = (ev: React.KeyboardEvent<HTMLElement>) => {
+        // eslint-disable-next-line default-case
+        switch (ev.keyCode) {
```

Comment:
I personally avoid using switch, I find them unnecessarily verbose and not performant, usually I use an object as a map, in this case it would look like:

```jsx
const KeyToHandler = {
    [KEY_CODE_MAP.ESC]: handleClose,
    [KEY_CODE_MAP.ENTER]: handleClick,
    [KEY_CODE_MAP.SPACE]: handleClick,
}

const handler = KeyToHandler[ev.keyCode]

if (handler) handler()
```

Consider changing it if you like the approach

Reply:
this doesn't work tho, have you tested this actually on component, or this was just suggested code?

Also, regarding performance and verbosity, object as a lookup map is not **always** the best option, following are reasons why switch may actually be better here:

1. _Different cases that return same value/code can be mapped together by placing case one after the other._ (which also means I can remove further 2 lines of code, thanks for making me realise that 😆 )
<img width="187" alt="Screenshot 2022-05-30 at 13 57 44" src="https://user-images.githubusercontent.com/14894844/170987651-38fcbf50-d5eb-4994-b661-486744fb7a46.png">

2. We can also control the order of cases for checking. 

3. **Object takes memory**. While using **switch**, or **if...else** does not require any additional memory. -> which we need to consider for in this case, as we won't be triggering this method that often, that it would be, in my opinion, worth it to define object literal for just 2 different options. And for method that will be triggered very scarcely.4. 

4. Does it work faster using Object literal? (taken from article linked below)
<img width="773" alt="Screenshot 2022-05-30 at 14 08 02" src="https://user-images.githubusercontent.com/14894844/170989232-6f680157-9ecd-4451-8364-7b506e353ab2.png">


I don't wanna write too much, but I would recommend this article: https://medium.com/front-end-weekly/switch-case-if-else-or-a-lookup-map-a-study-case-de1c801d944 :)

Reply:
I didn't test the code, it was just a suggestion, but I find it strange that it does not work.
I understand that using switch for smaller values can have even better performance, but I (personally) don't like the syntax, anyway it was just a suggestion, you can totally leave the switch.